src/rgw/rgw_rest_swift.cc

```diff
@@ -914,6 +914,109 @@ void RGWOptionsCORS_ObjStore_SWIFT::send_response()
   end_header(s, NULL);
 }
 
+int RGWBulkDelete_ObjStore_SWIFT::get_data(list<RGWBulkDelete::acct_path_t>& items,
+                                           bool * const is_truncated)
+{
+  const size_t MAX_LINE_SIZE = 2048;
+
+  RGWClientIOStreamBuf ciosb(*s->cio, (size_t)s->cct->_conf->rgw_max_chunk_size);
+  istream cioin(&ciosb);
+
+  char buf[MAX_LINE_SIZE];
+  while (cioin.getline(buf, sizeof(buf))) {
+    string path_str(buf);
+
+    ldout(s->cct, 20) << "extracted Bulk Delete entry: " << path_str << dendl;
+
+    RGWBulkDelete::acct_path_t path;
+
+    const size_t sep_pos = path_str.find('/');
+    if (string::npos == sep_pos) {
+      url_decode(path_str, path.bucket_name);
+    } else {
+      string bucket_name;
+      url_decode(path_str.substr(0, sep_pos), bucket_name);
+
+      string obj_name;
+      url_decode(path_str.substr(sep_pos + 1), obj_name);
+
+      path.bucket_name = bucket_name;
+      path.obj_key = obj_name;
+    }
+
+    items.push_back(path);
+
+    if (items.size() == MAX_CHUNK_ENTRIES) {
+      *is_truncated = true;
+      return 0;
+    }
+  }
+
+  *is_truncated = false;
+  return 0;
+}
+
+void RGWBulkDelete_ObjStore_SWIFT::send_response()
+{
+  set_req_state_err(s, ret);
+  dump_errno(s);
+  end_header(s, NULL);
+
+  s->formatter->open_object_section("delete");
+
+  const auto num_deleted = deleter->get_num_deleted();
+  const auto num_unfound = deleter->get_num_unfound();
+  const auto& failures = deleter->get_failures();
+
+  string resp_status;
+  string resp_body;
+
+  if (!failures.empty()) {
+    int reason = ERR_INVALID_REQUEST;
+    for (const auto fail_desc : failures) {
+      if (-ENOENT != fail_desc.err && -EACCES != fail_desc.err) {
+        reason = fail_desc.err;
+      }
+    }
+
+    rgw_err err;
+    set_req_state_err(err, reason, s->prot_flags);
+    dump_errno(err, resp_status);
+  } else if (0 == num_deleted && 0 == num_unfound) {
+    /* 400 Bad Request */
+    dump_errno(400, resp_status);
+    resp_body = "Invalid bulk delete.";
+  } else {
```

Comment:
Swift returns ```200 OK``` in case where all failures were due to lack of a given entity or access to it.

```
$ curl -i "$publicURL?bulk-delete" -H "X-Auth-Token: $token" -H "Content-Type: text/plain" -d "`cat /tmp/del`" -H "Accept: text/plain"
HTTP/1.1 200 OK
Content-Type: text/plain
X-Trans-Id: tx591c840467564bb0bb97b-005630ea1f
Date: Wed, 28 Oct 2015 15:30:39 GMT
Transfer-Encoding: chunked

Number Deleted: 0
Number Not Found: 6
Response Body: 
Response Status: 200 OK
Errors:
```